src/kernel.rs

```diff
@@ -52,34 +43,31 @@ impl EraInfo {
         self.era_id
     }
 
+    #[allow(clippy::result_large_err)]
     pub fn validate(
         &self,
-        block_header_with_signatures: &BlockHeaderWithSignatures,
+        JsonBlockWithSignatures { block, proofs }: JsonBlockWithSignatures,
     ) -> Result<(), BlockSignaturesValidationError> {
         let mut block_signature_weight = U512::from(0);
-        let block_header = block_header_with_signatures.block_header();
-        if block_header.era_id() != self.era_id {
+        if block.era_id() != self.era_id {
             return Err(BlockSignaturesValidationError::WrongEraId {
                 trusted_era_id: self.era_id,
-                block_header_era_id: block_header_with_signatures.block_header().era_id(),
+                block_header_era_id: block.era_id(),
             });
         }
         // See: https://github.com/casper-network/casper-node/blob/8ca9001dabba0dae95f92ad8c54eddd163200b5d/node/src/types/block.rs#L2465-L2474
-        let mut signature_data = block_header.block_hash().as_ref().to_vec();
-        signature_data.extend_from_slice(&block_header.era_id().to_le_bytes());
-        for (public_key, signature) in block_header_with_signatures
-            .block_signatures()
-            .proofs()
-            .iter()
-        {
+        let mut signature_data = block.hash().inner().into_vec();
```

Comment:
So the `hash` on a block could possibly be forged (this accessor doesn't compute the hash, it access it from the `Block` data structure).

You need to add a check that `block.block_header().block_hash() == block.hash()`

Reply:
You would also want to check the hash of the block body to make sure it matches the body hash in the header.

Reply:
Fixed